src/ssh-generator/ssh-generator.c

```diff
@@ -0,0 +1,476 @@
+#include "special.h"
+#include "virt.h"
+
+/* A small generator binding potentially five or more SSH sockets:
+ *
+ *     1. Listen on AF_VSOCK port 22 if we run in a VM with AF_VSOCK enabled
+ *     2. Listen on AF_UNIX socket /run/host/unix-export/ssh if we run in a container with /run/host/ support
+ *     3. Listen on AF_UNIX socket /run/ssh-unix-local/socket (always)
+ *     4. Listen on any socket specified via kernel command line option systemd.ssh_listen=
+ *     5. Similar, but from system credential ssh.listen
+ *
+ * The first two provide a nice way for hosts to connect to containers and VMs they invoke via the usual SSH
+ * logic, but without waiting for networking or suchlike. The third allows the same for local clients. */
+
+static const char *arg_dest = NULL;
+static bool arg_auto = true;
+static char **arg_listen_extra = NULL;
+
+static int parse_proc_cmdline_item(const char *key, const char *value, void *data) {
+        int r;
+
+        assert(key);
+
+        if (proc_cmdline_key_streq(key, "systemd.ssh_auto")) {
+                r = value ? parse_boolean(value) : 1;
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse systemd.ssh_auto switch \"%s\", ignoring: %m", value);
+                else
+                        arg_auto = r;
+
+        } else if (proc_cmdline_key_streq(key, "systemd.ssh_listen")) {
+
+                if (proc_cmdline_value_missing(key, value))
+                        return 0;
+
+                SocketAddress sa;
+                r = socket_address_parse(&sa, value);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse systemd.ssh_listen= expression, ignoring: %s", value);
+                else {
+                        _cleanup_free_ char *s = NULL;
+                        r = socket_address_print(&sa, &s);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to format socket address: %m");
+
+                        if (strv_consume(&arg_listen_extra, TAKE_PTR(s)) < 0)
+                                return log_oom();
+                }
+        }
+
+        return 0;
+}
+
+static int make_sshd_template_unit(
+                const char *dest,
+                const char *template,
+                const char *sshd_binary,
+                const char *found_sshd_template_service,
+                char **generated_sshd_template_unit) {
+
+        int r;
+
+        assert(dest);
+        assert(template);
+        assert(sshd_binary);
+        assert(generated_sshd_template_unit);
+
+        /* If the system has a suitable template already, symlink it to the name we want to reuse it */
+        if (found_sshd_template_service)
+                return generator_add_symlink(
+                                dest,
+                                template,
+                                /* dep_type= */ NULL,
+                                found_sshd_template_service);
```

Comment:
Shouldn't we just ship these templates ourselves? If we prefix them with `systemd-` it should be totally fine? And then drop the logic here to generate these ourselves.

Reply:
sshd can be anywhere, we search for it above. And I figured it would be nice to not leak ssh stuff into systems that don't have sshd installed at all. This way, if you don't have sshd there's really *nothing* sshd like in the system

Reply:
Isn't his better to solve at a packaging level? Fedora would have a systemd-ssh package which isn't installed by default but is a `Recommends` dependency of the sshd package?

I would leave it up to the distributions to decide how they want to handle this.

Reply:
the generator is so tiny, I am not sure it's worth splitting this off..

i mean, if people want to split it off they can, but i personally would just be annoyed if you split every tiny thing off. i mean, texlive, fedora, yuck.

Reply:
it's kind-a easier to investigate units when they are normal files on a non-live system (potentially from a package archive or a mounted system image).

Reply:
Also, don't we already ship the userdb ssh dropin unconditionally as well? So we're already dropping ssh specific files, so a few more won't hurt?

I don't really care too much which of the two approaches we use to be honest, but both seems messy. I would either always generate or always use templates in which case we ship them ourselves.

One thing that could potentially help is to put the ssh templates in some non-default lookup path and only have the generator put them in a default lookup path if sshd is found.

Reply:
so, here's the thing: we don't actually generate the service units at all in sane distros, because we simply reuse the distro provided ones. We only generate our own sshd service unit if we notice sshd is there, but the distro provides no sshd@.service unit.

Why that? Simply because at this point Fedora ships sshd@.service but suse does not, it only ships a sshd.service (i.e. no socket activation), and that for some political reasons afaik.

So in a way you actually have the behaviour you want: we'll just use distro-provided sshd service files if distros are up for it, and do not make up our own. We only make up our own as a fallback.

Hence, yes, I think it would be great if all distros would follow Fedora's lead there ans shop socket activation sshd@.service files, but this doesn't appear to be easily coming, or at all.

Reply:
@poettering Right, but you have no idea whether the distro shipped sshd template will actually do what you want. It could be templated on anything. Crazy enough people might have sshd templated on the port to listen on. So us introducing this assumption that if sshd@.service exists it is fit to be used as a xinetd service template seems rather dangerous.

That's why I prefer we ship this ourselves so that we're 100% sure that the template does what we need it to do and aren't relying on distributions shipping a sane sshd@ template.

Reply:
uh, i think people would be super pissed if we systematically bypass the distro shipped units.

At least on fedora the sshd@.service unit is entirely fine.

I think we should whenever we can use the distro stuff, if they ship that. At least at this point I am not aware of any distro that would ship a templated sshd@.service but would use it for anything else but socket activation.

I'd really just try it like this, and if we get feedback that things really don't work we can revisit this and bypass distros. But it's kind of a strong move to just bypass the distros even if they actually set things right for us.